### PR TITLE
Ninja changes

### PR DESCRIPTION
**Pull request by printer16**
//devices/uplink.dm (obj/items one) ~ New Ninja uplink added.
//datums/uplink_categories.dm ~ New category (Ninja modules) 
//ninja_modules.dm ~ New category (Ninja modules)
//outsider/ninja.dm ~ Ninja's spawn with the Ninja uplink.
//modules/ninja.dm ~ new modules including a frag launcher, matter fab, EMAG hand, EMP dissapation, emergency power generator, etc. 
//modules/computer.dm ~module moved and placed into ninja.dm
//light.dm ~ Ninja hardsuit boots now silent + slightly buffed to be bullet resistant
//all of the uplink datums. Changed prices in the ninja modules uplink. Made them restricted based on antag roles. #1992 should be fixed before merging for this to work properly.

### CODE REVIEW

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 1]
```diff
@@ -205,3 +208,108 @@
...
+		H << "<span class='danger'>Your suit emits a loud sound as power is rapidly injected into your suits battery!</span>"
+		playsound(H.loc, 'sound/effects/sparks2.ogg', 50, 1)
+		holder.cell.give(generation_ammount)
+		cooldown = 1
+		spawn(240)
+			cooldown = 0
+
+/obj/item/rig_module/emag_hand
+	name = "EMAG integrated hand."
+	desc = "A complex uprade that allows the user to touch things with their hand and apply an EMAG effect. High power cost."
```
Comment (Lohikar) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: This name and desc could be improved, remember that emags are not called emags in-game.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 2]
```diff
@@ -205,3 +208,108 @@
...
+
+	// Are we close enough?
+	var/mob/living/carbon/human/H = holder.wearer
+	if(!target.Adjacent(H))
+		return 0
+
+	H << "<span class = 'danger'>You stick your hand on [target] shorting out some of its circuts!</span>"
+	interfaced_with = target
+	target.emag_act(user, src)
+	spawn(2)
```
Comment (Lohikar) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Bad spawn/indentation here.

File: code/datums/uplink/ninja_modules.dm [hunk 3]
```diff
@@ -63,8 +63,8 @@
 	path = /obj/item/rig_module/power_sink
 
 /datum/uplink_item/item/ninja_modules/emag_hand
-	name = "EMAG Hand Module"
-	desc = "A module that allows the user to apply an EMAG effect to the targeted item."
+	name = "Electromagnetic Scrambler"
```
Comment (Lohikar) [inline code/datums/uplink/ninja_modules.dm]: This should probably be lowercase as well.

**Comment by printer16** [comment]
Changed around the capitalization on the other items as well. All lowercase now.

File: code/datums/uplink/uplink_categories.dm [hunk 4]
```diff
@@ -1,49 +1,117 @@
 /datum/uplink_category
 	var/name = ""
 	var/list/datum/uplink_item/items
+	var/list/datum/antagonist/antag_roles  // added this so category wide restrictions can be placed to make it eaiser to set restrictions
 
 /datum/uplink_category/New()
 	..()
 	items = list()
+	antag_roles = list()
```
Comment (skull132) [inline code/datums/uplink/uplink_categories.dm]: This is a dumb way to do it. Even though the way it generates items is also dumb, it shouldn't be perpetuated. You can just assign `var/list/datum/antagonist/antag_roles = list()` in the prototype. Same for all of the subclasses, there's no need to write out the New() proc for them.

Unless there's an empirical need. Though, I don't see any.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 5]
```diff
@@ -205,3 +208,107 @@
...
+
+	//Target wasn't supplied or we're already draining.
+	if(interfaced_with)
+		return 0
+
+	if(!target)
+		return 1
+
+	// Are we close enough?
+	var/mob/living/carbon/human/H = holder.wearer
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: H is not necessary here. You do no interactions that require typecasting to carbon/human, you can just keep referencing `holder.wearer`. I mean, I guess H is purdier, but eh.

File: code/datums/uplink/uplink_categories.dm (hunk 4 above)
Comment (printer16) [inline code/datums/uplink/uplink_categories.dm]: Well, my concern was you can see blank categories. This is a special case where an antag type is introduced that can only see one category. I tried to ensure that having categories be given the New() proc was a last resort. I tried placing the antag roles under the prototype and the items within and it didn't work. If you are okay with players being able to see blank categories, then I can remove it and instead assign it to the prototype item, but personally I don't like being able to see categories that I can't buy from.

**Comment by printer16** [comment]
That should do. Keeping them in the categories so people can't see blank categories as it looks ugly.

File: code/datums/uplink/ninja_modules.dm [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/***************
+* Special Ninja Modules *
+***************/
+/datum/uplink_item/item/ninja_modules
+	category = /datum/uplink_category/ninja_modules
+
+/datum/uplink_item/item/ninja_modules/self_destruct
+	name = "Self Destruct Module"
```
Comment (skull132) [inline code/datums/uplink/ninja_modules.dm]: Self-Destruct

File: code/game/objects/items/devices/uplink.dm [hunk 7]
```diff
@@ -370,3 +370,25 @@ A list of items and costs is stored under the datum of every game mode, alongsid
 /obj/item/device/contract_uplink/attack_self(mob/user as mob)
 	if (hidden_uplink)
 		hidden_uplink.trigger(user)
+
+/*
+ * A ninja uplink
+ */
+
+/obj/item/device/ninja_uplink
+	name = " modified contract uplink"
```
Comment (skull132) [inline code/game/objects/items/devices/uplink.dm]: Spacebar right after the opening quotes. Remove it.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 8]
```diff
@@ -205,3 +208,107 @@
...
+	holder.emp_protection += protection_amount
+
+/obj/item/rig_module/emp_shielding/deactivate()
+	if(!..())
+		return
+
+	holder.emp_protection = max(0,(holder.emp_protection - protection_amount))
+
+/obj/item/rig_module/emergency_powergenerator
+	name = "Emergency Power Generator"
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Inconsistent naming in the file. Lower the letters on this.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 9]
```diff
@@ -205,3 +208,107 @@
...
+	name = "EMP dissipation module"
+	desc = "A bewilderingly complex bundle of fiber optics and chips. Seems like it uses a good deal of power."
+	active_power_cost = 10
+	toggleable = 1
+	usable = 0
+
+	activate_string = "Enable active EMP shielding"
+	deactivate_string = "Disable active EMP shielding"
+
+	interface_name = "active EMP shielding system"
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Upper first, lower all but acronym.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 10]
```diff
@@ -205,3 +208,107 @@
...
+/obj/item/rig_module/emergency_powergenerator
+	name = "Emergency Power Generator"
+	desc = "A high yield power generating device that takes a long time to recharge."
+	active_power_cost = 0
+	toggleable = 0
+	usable = 1
+	confined_use = 1
+	var/cooldown = 0
+
+	engage_string = "Use Emergency Power"
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Inconsistent naming on engage_string throughout the file. (Capitalization.)

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 11]
```diff
@@ -205,3 +208,107 @@
...
+
+	interface_name = "Emergency Power Generator"
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Inconsistent naming of interface_name throughout the file. (Capitalization.)

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 12]
```diff
@@ -205,3 +208,107 @@
...
+	else if(cooldown == 0)
+		H << "<span class='danger'>Your suit emits a loud sound as power is rapidly injected into your suits battery!</span>"
+		playsound(H.loc, 'sound/effects/sparks2.ogg', 50, 1)
+		holder.cell.give(generation_ammount)
+		cooldown = 1
+		spawn(240)
+			cooldown = 0
+
+/obj/item/rig_module/emag_hand
+	name = "EMAG integrated hand."
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Remove the period from the end here.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 13]
```diff
@@ -205,3 +208,107 @@
...
+		return 1
+
+	// Are we close enough?
+	if(!target.Adjacent(holder.wearer))
+		return 0
+
+	holder.wearer << "<span class = 'danger'>You stick your hand on [target] shorting out some of its circuts!</span>"
+	interfaced_with = target
+	target.emag_act(user, src)
+	spawn(2)
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Statement doesn't do anything. Spawn() applies to everything tabbed in from it.
```
spawn(n)
    this_waits()
this_continues_immediately()
```

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm [hunk 14]
```diff
@@ -205,3 +208,107 @@
...
+	var/mob/living/carbon/human/H = holder.wearer
+	if(cooldown == 1)
+		H << "<span class='danger'>There isn't enough power stored up yet!</span>"
+		return 0
+	else if(cooldown == 0)
+		H << "<span class='danger'>Your suit emits a loud sound as power is rapidly injected into your suits battery!</span>"
+		playsound(H.loc, 'sound/effects/sparks2.ogg', 50, 1)
+		holder.cell.give(generation_ammount)
+		cooldown = 1
+		spawn(240)
```
Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Consider using:
```
schedule(src, /obj/item/rig_module/emergency_powergenerator/proc/reset_cooldown, 240)
```
With reset_cooldown as:
```
/obj/item/rig_module/emergency_powergenerator/proc/reset_cooldown()
    cooldown = 0
```

Comment (Lohikar) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: `schedule(src, .proc/reset_cooldown, 240)` should also work, if you want something a bit shorter.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm (hunk 13 above)
Comment (Lohikar) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: Circuits is also misspelled above.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm (hunk 10 above)
Comment (printer16) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: First letter of every new word is capitalized with engage_string.

File: code/modules/clothing/spacesuits/rig/modules/ninja.dm (hunk 9 above)
Comment (printer16) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: If you mean the activate_string all of the other items that were in there had the first letter of every new word capitalized.

Comment (skull132) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: I mean the line that I commented on. Pretty sure that interface_name is supposed to have the first letter capitalized.

Comment (printer16) [inline code/modules/clothing/spacesuits/rig/modules/ninja.dm]: The others don't.

**Comment by OneOneThreeEight** [comment]
Why does the ninja suit have a mounted frag grenade launcher?

**Comment by printer16** [comment]
@OneOneThreeEight There is an option to buy it. They don't start with it. It costs over half of their allotted TC's and only has 3 grenades that can't be refilled without having more frag grenades. Which they can't buy. And frag grenades are hard to get. So if they do pick this they are sacrificing the ability to get longer term modules.

**Pull request closed without merging**